iceoryx_hoofs/include/iceoryx_hoofs/design_pattern/static_lifetime_guard.hpp

```diff
@@ -0,0 +1,117 @@
+// Copyright (c) 2022 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#ifndef IOX_HOOFS_DESIGN_STATIC_LIFETIME_GUARD_HPP
+#define IOX_HOOFS_DESIGN_STATIC_LIFETIME_GUARD_HPP
+
+#include <atomic>
+#include <type_traits>
+
+namespace iox
+{
+namespace design_pattern
+{
+/// @brief Manages a static instance of type T in a way so that each
+/// existing StaticLifetimeGuard prevents the destruction of
+/// the instance.
+/// 1. instance() creates a static guard itself and hence has static lifetime
+/// 2. any static StaticLifetimeGuard G created before that prolongs the lifetime
+/// of the instance at least until G is destroyed
+/// @tparam T the type of the instance to be guarded
+///
+/// @note dtor, ctor and copy ctor and instance are thread-safe
```

Comment:
Why not move and setCount?

Reply:
move is at well. All functions are so, I will mark the entire class as thread-safe.

Reply:
Added more explanation. Also `setCount` is a (necessary?) evil for testing at the moment and can be dangerous if used incorrectly. Worst case is no destruction or destruction after main but earlier than expected. This is tolerable and hard to avoid completely (non-guarded instances have worse problems IMO).